python/pyarrow/tensorflow/plasma_op.cc

```diff
@@ -0,0 +1,383 @@
+// specific language governing permissions and limitations
+// under the License.
+
+#include "tensorflow/core/framework/device_base.h"
+#include "tensorflow/core/framework/op.h"
+#include "tensorflow/core/framework/op_kernel.h"
+#include "tensorflow/core/framework/shape_inference.h"
+#include "tensorflow/core/platform/logging.h"
+#include "tensorflow/core/platform/mutex.h"
+#include "tensorflow/stream_executor/device_memory.h"
+#include "tensorflow/stream_executor/event.h"
+#include "tensorflow/stream_executor/stream.h"
+
+#ifdef GOOGLE_CUDA
+#include "tensorflow/core/common_runtime/gpu/gpu_event_mgr.h"
+#include "tensorflow/core/platform/stream_executor.h"
+#endif
+
+#include "arrow/adapters/tensorflow/convert.h"
+#include "arrow/io/memory.h"
+#include "arrow/python/arrow_to_python.h"
+#include "arrow/python/python_to_arrow.h"
+#include "arrow/tensor.h"
+#include "plasma/client.h"
+
+
+namespace tf = tensorflow;
+
+using ArrowStatus = arrow::Status;
+using CPUDevice = Eigen::ThreadPoolDevice;
+using GPUDevice = Eigen::GpuDevice;
+
+using Event = perftools::gputools::Event;
+using Stream = perftools::gputools::Stream;
+
+// NOTE(zongheng): for some reason using unique_ptr or shared_ptr results in
+// CUDA_ERROR_DEINITIALIZED on program exit.  I suspect this is because the
+// static object's dtor gets called *after* TensorFlow's own CUDA cleanup.
+// Instead, we use a raw pointer here and manually clean up in the Ops' dtors.
+static Stream* d2h_stream = nullptr;
+static tf::mutex d2h_stream_mu;
+
+// TODO(zongheng): CPU kernels' std::memcpy might be able to be sped up by
+// parallelization.
+
+int64_t get_byte_width(const arrow::DataType& dtype) {
+  return arrow::checked_cast<const arrow::FixedWidthType&>(dtype).bit_width() / CHAR_BIT;
+}
+
+// Put:  tf.Tensor -> plasma.
+template <typename Device>
+class TensorToPlasmaOp : public tf::AsyncOpKernel {
+ public:
+  explicit TensorToPlasmaOp(tf::OpKernelConstruction* context) : tf::AsyncOpKernel(context) {
+    OP_REQUIRES_OK(context, context->GetAttr("plasma_store_socket_name",
+                                             &plasma_store_socket_name_));
+    OP_REQUIRES_OK(context, context->GetAttr("plasma_manager_socket_name",
+                                             &plasma_manager_socket_name_));
+    tf::mutex_lock lock(mu_);
+    if (!connected_) {
+      VLOG(1) << "Connecting to Plasma...";
+      ARROW_CHECK_OK(client_.Connect(plasma_store_socket_name_,
+                                     plasma_manager_socket_name_,
+                                     plasma::kPlasmaDefaultReleaseDelay));
+      VLOG(1) << "Connected!";
+      connected_ = true;
+    }
+  }
+
+  ~TensorToPlasmaOp() override {
+    {
+      tf::mutex_lock lock(mu_);
+      ARROW_CHECK_OK(client_.Disconnect());
+      connected_ = false;
+    }
+    {
+      tf::mutex_lock lock(d2h_stream_mu);
+      if (d2h_stream != nullptr) {
+        delete d2h_stream;
+        d2h_stream = nullptr;
```

Comment:
(Communicated offline as well, copy pasting here)

I don’t understand this fix. Why would the destructor of the same object be called twice?
Can you (1) revert this commit, (2) change the Python test from `sess = tf.Session(...); ...` to `with tf.Session(...) as sess:`?  Just want to see if this is some default graph semantics or something like that (likely not)
Another independent change to try (1) revert this commit, (2) add `.SetIsStateful()` to the op’s registration, the two REGISTER_OP() in memcpy_plasma_op.cc